lib/usb-device.js

```diff
@@ -0,0 +1,210 @@
+    let endpoints = this.deviceInterface.endpoints;
+    // set the endpoints for our USBDevice object
+    // note that these endpoints are on the device, so it's endpoint in
+    // is our endpoint out, the endpoint we want to send data to
+    if (endpoints[0] === LIBUSB_ENDPOINT_IN) {
+      this.inputEndpoint = endpoints[1];
+      this.outputEndpoint = endpoints[0];
+    } else {
+      this.inputEndpoint = endpoints[0];
+      this.outputEndpoint = endpoints[1];
+    }
+    this.inputEndpoint.transferType = LIBUSB_TRANSFER_TYPE_BULK;
+    this.outputEndpoint.transferType = LIBUSB_TRANSFER_TYPE_BULK;
+    this.inputEndpoint.timeout = 2500;
+    this.outputEndpoint = Promise.promisifyAll(this.outputEndpoint);
+    this.inputEndpoint = Promise.promisifyAll(this.inputEndpoint);
+  }
+
+  // handles authentication between our machine and
+  // the android device upon initial connection
+  // TODO: attempt to sign token sent to us
+  async handleAuth () {
+    let publicKeyBuf = new Buffer(publicKeyString.length + 1);
+    publicKeyString.copy(publicKeyBuf);
+    publicKeyBuf[-1] = 0;
+    let authMsg = generateMessage(ADB_COMMANDS.CMD_AUTH,
+                                  3, 0, publicKeyBuf);
+
+    let command = null;
+    do {
+      await this._sendMsg(authMsg);
+      await this._sendMsg(publicKeyBuf);
+      console.log("Sent auth message and public key.");
+      let packet = await this._recvMsg(MAXDATA);
+      //TODO, actually check the fields of the packet
+      if (packet.command === ADB_COMMANDS.CMD_CNXN) {
+        // we need to read the string about the device that gets sent back
+        // await this.inputEndpoint.transferAsync(MAXDATA);
+        return;
+      }
+    } while (command !== ADB_COMMANDS.CMD_CNXN);
+  }
+
+  // calls claim device and then performs the ADB usb handshake
+  // returns a promise, will have the error value if an error is thrown
+  async initConnection () {
+    await this.claimDevice();
+    console.log("Trying to connect to device.");
+    let connectMsg = generateMessage(ADB_COMMANDS.CMD_CNXN
+                                    , CONNECT_VALUES.CONNECT_VERSION
+                                    , CONNECT_VALUES.CONNECT_MAXDATA
+                                    , CONNECT_VALUES.CONNECT_PAYLOAD);
+    let hostBuf = new Buffer(CONNECT_VALUES.CONNECT_PAYLOAD);
+    await this._sendMsg(connectMsg);
+    await this._sendMsg(hostBuf);
+    console.log("Sent connect message.");
+    let packet = await this._recvMsg(MAXDATA);
+    if (packet.command === ADB_COMMANDS.CMD_AUTH) {
+      console.log("AUTH received.");
+      await this.handleAuth();
+    }
+  }
+
+  // sends a message, checks that message is a Buffer
+  async _sendMsg (msg) {
+    // this error checking breaks the function :/
+    // if(typeof msg !== Buffer) {
+    //   throw new Error("Message was not of type Buffer.");
+    // }
+    return await this.outputEndpoint.transferAsync(msg);
```

Comment:
`return await ` is actually the same as just `return`

in the first, we get a value from a promise and wrap it with a promise. in the second, we return the promise directly. so we've been preferring the second method